stingray/pulse/search.py

```diff
@@ -549,3 +553,106 @@ def phaseogram(
         additional_info = {}
 
     return phas, binx, biny, additional_info
+
+
+class Pulsar_Periodogram:
+    """
+    Creates a Periodogram for a given event
+
+    Parameters
+    ----------
+    events : EventList
+        The input Event list to make the periodogram
+    fmin   : float
+        The minimum frequency which is evaluated
+    fmax   : float
+        The maximum frequency which is evaluated
+    oversample : int
+        Factor for improved resolution
+    nharm  : int
+        Number of harmonics
+    nbin : int
+        The number of bins/ degree of freedom
+
+    Attributes
+    ----------
+    obs_length: float
+        Total time of the observation
+
+    frequencies: np.array
+        Array of the frequencies to be checked
+
+    zstat: np.array
+        Z statistics of the respective frequencies
+
+    z_detlev: float
+        The level above which a frequency will be
+        considered as a candidate pulse
+
+    cand_freq_z: np.array
+        Candidate frequencies of the periodogram
+    """
+
+    def __init__(self, events: EventList, fmax, fmin, oversample=1, nharm=1, nbin=32):
+        self.events = events
+        self.fmax = fmax
+        self.fmin = fmin
+        self.oversample = oversample
+        self.nharm = nharm
+        self.nbin = nbin
+        self.obs_length = events.time.max() - events.time.min()
+        self.df = 1 / (self.oversample * self.obs_length)
+        self.frequencies = np.arange(fmin, fmax, self.df)
+        self.zstat = z_n_search(events.time, self.frequencies, nbin=nbin, nharm=self.nharm)[1]
```

Comment:
The search should not be executed directly, I think. It would be better to allocate the periodogram and then calculate the zstat attribute the first time one tries to access it. One might even heve different attributes for z1, z2, z3, etc. corresponding to $Z^2_1$, $Z^2_2$, $Z^2_3$, etc.

Reply:
If you want a way to initialize the attributes only when needed, look at how `Lightcurve.countrate` is defined as a property, for example